[TypeScript & Module Typing] 위니(김예지) 미션 제출합니다.

rladpwl0512 opened the pull request:
안녕하세요~ 리뷰 잘 부탁드립니다 우연 😄 

## 🔥 팀 미션 적용 결과

현재 팀에서 유틸 타입을 따로 분리하지 않았습니다. 페어와 함께 유틸 타입으로 사용할만한 것이 없을지 찾았지만, 찾지 못했어요. 
여기서 저희가 생각한 유틸 타입이란, 우리 프로젝트 뿐 아니고 다른 사람들이 사용했을 때도 의미가 있는 범용적인 타입이라고 생각했습니다. 
물론 억지로 만들 수도 있었지만 근거 없이 만들지 말라고 하셨던것이 생각나서 일단 분리 하지 않았습니다. 하지만 공부할 겸, 작은 것이라도 유틸 타입을 만들고 출시해 볼 예정입니다. 

- [ ] 실제 코드 반영 후 팀 프로젝트에서 타입 테스트 진행
- [ ] 자주 사용되는 타입을 유틸로 출시하여 분리

## ✅ 요구사항 구현 확인

> 필히 테스트 & 타입이 포함

### DOM 유틸

- [x] innerHTML()
- [x] show()
- [x] hidden()
- [x] addEvent()

### 유틸

- [x] fetch()
- [x] isNull()
- [x] isNil()
- [x] isNumber()
- [x] isFunction()
- [x] shuffle()
- [x] pick()
- [x] omit()
- [x] memoize()
- [x] debounce()
- [x] throttle()
- [x] clickOutside()

## 🧐 공유
- 구현을 마지막으로 작성하는 과정에서, 이전에 선언해놓았던 타입을 변경 하게 되는 일이 많았습니다. 이러한 상황을 줄이기 위해서 어떻게 해야할지 고민되네요. 
타입선언 및 테스트 과정에서는 구현은 하지 않지만 함수가 무엇을 하는것인지에 대한 정의를 최대한 꼼꼼하게 하는것만이 방법일까요😅 
- 사실 아직 tsd 테스트에 대한 효용성을 느끼지 못하고 있습니다. 혹시 느낀 부분이 있으실까요? 
- 계속해서 타입 선언->타입 테스트를 진행하다가 이후에는 타입 테스트->타입 선언을 해주었습니다. 이유는 타입 테스트에서 다양한 케이스를 생각하고 그에 예상되는 타입을 지정해주면서 구현 전에 함수에 대한 정의를 더 구체적으로 할 수 있다는 느낌을 받았습니다. 우연은 어떤 방식으로 진행하셨는지 궁금하네요! 

혹시 궁금한 점 있거나 빠진 부분이 있으면 바로 알려주세요~ 잘부탁드려요 😄

rladpwl0512 committed:
chore: tsconfig.json 초기설정

rladpwl0512 committed:
feat: isNull 타입 선언

rladpwl0512 committed:
test: isNull 테스트 작성

rladpwl0512 committed:
feat: isNil 타입 선언

rladpwl0512 committed:
test: isNil 테스트 작성

rladpwl0512 committed:
feat: isNumber 타입 선언

rladpwl0512 committed:
test: isNumber 테스트 작성

rladpwl0512 committed:
feat: isFunction 타입 선언

rladpwl0512 committed:
test: isFunction 테스트 작성

rladpwl0512 committed:
feat: shuffle 타입 선언

rladpwl0512 committed:
test: shuffle 테스트 작성

rladpwl0512 committed:
feat: pick 타입 선언

rladpwl0512 committed:
test: pick 테스트 작성

rladpwl0512 committed:
feat: omit 타입 선언

rladpwl0512 committed:
test: omit 테스트 작성

rladpwl0512 committed:
feat: debounce 타입 선언

rladpwl0512 committed:
test: debounce 테스트 작성

rladpwl0512 committed:
feat: throttle 타입 선언

rladpwl0512 committed:
test: throttle 테스트 작성

rladpwl0512 committed:
feat: memoize 타입 선언

rladpwl0512 committed:
test: memoize 테스트 작성

rladpwl0512 committed:
feat: fetch 타입 선언

rladpwl0512 committed:
test: fetch 테스트 작성

rladpwl0512 committed:
feat: declare 추가

rladpwl0512 committed:
test: clickOutside 테스트 작성

rladpwl0512 committed:
feat: clickOutside 타입 선언

rladpwl0512 committed:
chore: 기본 설정

rladpwl0512 committed:
feat: util 함수 구현

rladpwl0512 committed:
test: dom 유틸에 대한 테스트 작성

rladpwl0512 committed:
test: dom 유틸에 대한 테스트 수정

- HTMLElement 뿐 아니라 undefined가 들어올 수 있기 때문에 이에 대해 대응

rladpwl0512 committed:
feat: dom 유틸에 타입 선언 및 구현

rladpwl0512 committed:
test: memoize 테스트 작성

rladpwl0512 committed:
feat: memoize 타입 선언 및 구현

rladpwl0512 committed:
feat: index.d.ts 빌드

ronci commented on src/index.d.ts:
```diff
@@ -0,0 +1,25 @@
+    show: () => void;
+    addEvent: <K extends keyof HTMLElementEventMap>(type: K, listener: (this: HTMLElement, ev: HTMLElementEventMap[K]) => any) => void;
+}
+declare function _(selector: string): HTMLElement | undefined;
```
다음 type을 왜 `HTMLElement | undefined`로 정의하셨는 지 알 수 있을까요?

rladpwl0512 replied:
제가 직접 정의해준 것은 아니고, 구현부에 따라 타입스크립트가 추론을 해준 것 같아요! 
`element`가 있을 때는 `HTMLelment`를 리턴해주고, 없을 때는 `undefined`를 리턴해주기 때문에 이렇게 된 것 같습니다! 
추론이 제대로 되고있기 때문에, 따로 리턴 타입을 선언해주지 않았습니다!

ronci commented on src/index.ts:
```diff
@@ -1,45 +1,167 @@
+    input: RequestInfo | URL,
+    init?: RequestInit
+  ): Promise<Response> {
+    return fetch(input, init);
```
이렇게 사용했을 때 fetch가 중복되어 선언되어 있어서 문제는 따로 없었나요??

rladpwl0512 replied:
네 잘 인식하고 있더라구요! 그런데 어떻게 잘 인식하는지는 정확히 모르겠습니다 😅 
찾아봤는데, 명확한 답을 찾지는 못했네요 😅 
일단 코드를 읽는데 헷갈릴 것 같아서, 정의해 준 `fetch` 함수명을 `fetchData`로 변경했습니다.

ronci commented on src/index.ts:
```diff
@@ -1,45 +1,167 @@
+  }
+
+  // input이 null인지 확인하여 boolean값을 반환
+  export function isNull<T>(value: T): T extends null ? true : false;
+  export function isNull(value: unknown): boolean {
+    return value === null;
+  }
```
다음과 같이 type을 두 번 정의해주신 이유가 있으신가요??

rladpwl0512 replied:
<img width="726" alt="스크린샷 2022-09-29 오후 4 24 20" src="https://user-images.githubusercontent.com/52344833/192966920-b01d506d-d36b-49ac-911a-c6a2ed120683.png">

처음에는 위의 정의를 가지고 구현까지 해주었는데, 위와 같은 오류가 나더라구요. 
boolean이 true나 false보다 더 큰 범위여서 나는 오류로 보입니다. 
이렇게 해주면 d.ts에는 위 타입이 선언이 되고 구현부는 아래로 인식이 되어서 일단 타입을 가장 좁혀줄 수 있도록 위 타입선언을 추가해주었습니다.

ronci commented on src/index.ts:
```diff
@@ -1,45 +1,167 @@
+  }
+
+  // obj에 대한 keys만 뽑은 새로운 obj를 반환
+  export function pick<T extends Record<string, unknown>, K extends keyof T>(
```
Record<string, unknown>으로 선언하면 다음과 같은 문제가 있을 수도 있다고 하네요!
[타입스크립트 인터페이스를 임의의 객체에 할당하지 못하는 문제](https://velog.io/@giwan_dev/til-typescript-interface-object)

rladpwl0512 replied:
이 부분은 아직 잘 이해되지 않아서 적용하지 못했네요 😅 더 알아보겠습니다!

ronci reviewed (changes requested):
안녕하세요! 위니의 리뷰어 우연입니다!

미션 구현하시느라 수고하셨어요!

밑에 코멘트 몇 개 달아놨는데 한 번 확인해주시면 좋을 것 같아요! 😊

---
## 위니의 질문에 대해서
- 구현을 마지막으로 작성하는 과정에서, 이전에 선언해놓았던 타입을 변경 하게 되는 일이 많았습니다. 이러한 상황을 줄이기 위해서 어떻게 해야할지 고민되네요.타입선언 및 테스트 과정에서는 구현은 하지 않지만 함수가 무엇을 하는것인지에 대한 정의를 최대한 꼼꼼하게 하는것만이 방법일까요😅
    - 저는 미션을 해보면서 정의가 확실하게 세워지고 그것에 따라 함수가 input, output type이 확실하다면 타입을 변경할 일이 없다고 생각해요. 타입이 변경되었다는 것은 input이나 output 조건이 바껴서 함수의 사용자체가 달라졌다고 생각해서 위니가 말한대로 정의를 꼼꼼히 하는 것이 방법 아닐까 하네요!
- 사실 아직 tsd 테스트에 대한 효용성을 느끼지 못하고 있습니다. 혹시 느낀 부분이 있으실까요?
    - 이번에 팀프로젝트에서 타입스크립트 리팩터링을 해보면서 여러정보가 들어가는 api에 대해서 여러 곳에서 type을 pick하는 일이 있었어요. 지금은 바로 개발단계라 상관없겠지만 후에 봤을 때 로직만 남은 type에서 이 type이 어떤 것을 의미하는 지 잘 모르고 실수로 다른 곳의 type이 변경되었을 때 다른 곳에서는 type이 제대로 유지되는 지 확인할 수 있는 방법이 없더라고요. 그럴 때 tsd를 사용해서 “여기 A type에는 이 값들만 와야한다!”라고 테스트할 수 있으니까 그 때 쓰면 좋겠다는 생각이 들었어요!
- 계속해서 타입 선언->타입 테스트를 진행하다가 이후에는 타입 테스트->타입 선언을 해주었습니다. 이유는 타입 테스트에서 다양한 케이스를 생각하고 그에 예상되는 타입을 지정해주면서 구현 전에 함수에 대한 정의를 더 구체적으로 할 수 있다는 느낌을 받았습니다. 우연은 어떤 방식으로 진행하셨는지 궁금하네요!
    - 저는 타입선언 후 테스트로 진행하였는데요. 이유는 이 함수가 무슨일을 하는 지 제대로 이해를 하지못해서 type을 선언해보면서 어떻게 사용할지 생각했었어요. 그 후 테스트를 해보며 제대로 작동하는 지 확인하는 방법으로 진행하였습니다!
    - 하지만 지금 생각하면 첫 번째 질문에 대답한 것 처럼 함수 정의가 확실하게 되어있다면 타입 테스트를 한 후에 타입 선언 방식을 사용하겠네요! 그러면 위니가 말씀해주신대로 다양한 케이스에 맞춰서 내가 잘 선언을 하고 있는 지 확인할 수 있을 것 같아요!

rladpwl0512 committed:
refactor: 함수명 변경

fetch -> fetchData로 함수를 변경했습니다. 내부 로직에서 js에서 제공해주는 fetch를 리턴해주고 있기 때문에 헷갈릴 수도 있다고 판단되었습니다.

rladpwl0512 committed:
fix: memoize 유틸 함수에 대한 리턴 타입 수정

memoize는 () => void 형태의 함수를 리턴해주기 때문에, 이에 맞게 변경

rladpwl0512 committed:
chore: jest 속성을 jest 설정파일로 이동

rladpwl0512 committed:
fix: 작동하지 않는 test 수정

리턴타입이 타입 선언과 맞지 않아서 돌아가지 않는 테스트를 수정

rladpwl0512 committed:
feat: index.d.ts 빌드

rladpwl0512 commented:
우연~ 리뷰 꼼꼼하게 해주셔서 감사합니다! 커멘트 뿐 아니라, 제 고민에 대한 생각도 너무 자세히 작성해주셔서 많은 도움이 됐어요~ 
피드백 주신 부분들을 반영하고, 추가적으로 부족했던 부분들을 반영했습니다. 
시간이 부족해서 더 많은 것을 적용해보지 못한 것 같아서 아쉽네요 ... 😅 
일단 6시가 마감이기 때문에 맞춰서 제출합니다! 
아직 좀 부족한 부분들은 추가적으로 반영해보도록 하겠습니다!

ronci reviewed (approved):
위니 코드에 대해 잘 설명해주셔서 감사합니다! 고생하셨어요!
머지하겠습니다!
